File: python/Two Pointers/valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        """
        Check if a string is a palindrome, considering only alphanumeric characters and ignoring cases.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # Initialize two pointers: left at start, right at end
        l, r = 0, len(s) - 1
        
        # Continue until pointers meet in the middle
        while l < r:
            # Skip non-alphanumeric characters from left
            while l < r and not self.alphaNum(s[l]):
                l += 1
            # Skip non-alphanumeric characters from right
            while r > l and not self.alphaNum(s[r]):
                r -= 1

            # Compare characters (case-insensitive)
            if s[l].lower() != s[r].lower():
                return False
            # Move pointers towards center
            l, r = l + 1, r - 1
        return True    

    def alphaNum(self, c):
        # Check if character is alphanumeric using ASCII values
        return ((ord("A") <= ord(c) <= ord("Z")) or  # Uppercase letters
                (ord("0") <= ord(c) <= ord("9")) or  # Numbers
                (ord("a") <= ord(c) <= ord("z")))    # Lowercase letters

# -------------------
# Test Cases
# -------------------
def test_is_palindrome():
    s = Solution()
    # Test case 1: Example
    assert s.isPalindrome("A man, a plan, a canal: Panama") == True
    # Test case 2: Not a palindrome
    assert s.isPalindrome("race a car") == False
    # Test case 3: Empty string
    assert s.isPalindrome("") == True
    # Test case 4: Only non-alphanumeric
    assert s.isPalindrome("!!!") == True
    # Test case 5: Single character
    assert s.isPalindrome("a") == True
    # Test case 6: Numbers
    assert s.isPalindrome("12321") == True
    assert s.isPalindrome("1231") == False
    # Test case 7: Mixed case
    assert s.isPalindrome("AbBa") == True
    print("All test cases passed!")

if __name__ == "__main__":
    test_is_palindrome()

File: C++/Two Pointers/valid_palindrome.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Check if string is palindrome (alphanumeric chars only, case insensitive)
     * 
     * Approach:
     * - Two pointers: left at start, right at end
     * - Skip non-alphanumeric characters
     * - Compare characters case-insensitively
     * - Move towards center
     * 
     * Time: O(n) - single pass through string
     * Space: O(1) - only two pointers used
     */
    bool isPalindrome(string s) {
        int left = 0;
        int right = s.length() - 1;
        
        while (left < right) {
            // Skip non-alphanumeric from left
            while (left < right && !isAlphaNum(s[left])) {
                left++;
            }
            
            // Skip non-alphanumeric from right
            while (right > left && !isAlphaNum(s[right])) {
                right--;
            }
            
            // Compare characters case-insensitively
            if (tolower(s[left]) != tolower(s[right])) {
                return false;
            }
            
            left++;
            right--;
        }
        return true;
    }

private:
    bool isAlphaNum(char c) {
        return (isdigit(c) || isalpha(c));
    }
};

/*
 * Detailed C++ Syntax Explanation:
 * 
 * 1. String Operations:
 *    s.length()      // Get string length
 *    s[left]         // Access character at index
 *    tolower(c)      // Convert char to lowercase
 * 
 * 2. Character Functions:
 *    isdigit(c)      // Check if char is digit (0-9)
 *    isalpha(c)      // Check if char is letter (a-z,A-Z)
 * 
 * 3. Pointers/Indices:
 *    int left = 0    // Start of string
 *    int right = s.length() - 1  // End of string
 * 
 * 4. While Loop:
 *    while (left < right)  // Continue until pointers meet
 * 
 * 5. Private Helper:
 *    bool isAlphaNum(char c)  // Class member function
 *    - Returns true if c is letter or digit
 * 
 * Note: This implementation:
 * - Uses C++ standard library functions instead of ASCII values
 * - Has same O(n) time complexity as Python version
 * - Maintains O(1) space complexity
 */
